src/Control/Monad/Gen.purs

```diff
@@ -30,10 +29,29 @@ choose genA genB = chooseBool >>= if _ then genA else genB
 
 -- | Creates a generator that outputs a value chosen from a selection of
 -- | existing generators with uniform probability.
-oneOf :: forall m f a. MonadGen m => Foldable f => NonEmpty f (m a) -> m a
-oneOf (x :| xs) = do
-  n <- chooseInt 0 (length xs)
-  if n < 1 then x else fromIndex (n - 1) x xs
+oneOf :: forall m f a. MonadGen m => Foldable1 f => f (m a) -> m a
+oneOf xs = do
+  n <- chooseInt 0 (length xs - 1)
+  fromIndex n xs
+
+newtype FreqSemigroup a = FreqSemigroup (Number -> Tuple (Maybe Number) a)
+
+freqSemigroup :: forall a. Tuple Number a -> FreqSemigroup a
+freqSemigroup (Tuple weight x) =
+  FreqSemigroup \pos ->
+    if pos >= weight
+      then Tuple (Just (pos - weight)) x
+      else Tuple Nothing x
+
+getFreqVal :: forall a. FreqSemigroup a -> Number -> a
+getFreqVal (FreqSemigroup f) = snd <<< f
+
+instance semigroupFreqSemigroup :: Semigroup (FreqSemigroup a) where
+  append (FreqSemigroup f) (FreqSemigroup g) =
+    FreqSemigroup \pos ->
+      case f pos of
+           Tuple (Just pos') _ -> g pos'
+           result              -> result
```

Comment:
Nitpick: we prefer not to align code with whitespace (also even nittier: too much indentation for the cases :wink:)

Reply:
Will fix!